### PR TITLE
[TMTensor] Add lowering of attention with an explicit scale value

### PR DESCRIPTION
**Pull request by AGindinson**
This allows to pass along the scale constant when set at the Torch level and lifts the restriction on dynamic head dim size. This also drops strict validation of the attention value being equal to Torch SDPA's default of `1 / sqrt(headDim)` to accommodate more use-cases.

Essentially follows up on commit f53b5be.

### CODE REVIEW

File: lib/Conversion/TorchToTMTensor/TorchToTMTensor.cpp [hunk 1]
```diff
@@ -1934,32 +1935,31 @@ class ConvertAtenScaledDotProductAttentionOp
...
-      // 1e-6 relative tolerance is ~10x float32 machine epsilon, which
-      // provides a safe margin for:
-      // - Different computation orders (a*b vs b*a can differ slightly)
-      // - Float64 -> float32 -> float64 round-trips through serialization
-      double relativeError =
-          std::abs(scaleFloat - expectedScale) / expectedScale;
-      if (relativeError > 1e-6) {
-        return rewriter.notifyMatchFailure(
-            loc, "scale must be None or 1/sqrt(headDim)");
+        double expectedScale = 1.0 / std::sqrt(static_cast<double>(headDim));
```
Comment (ziereis) [inline lib/Conversion/TorchToTMTensor/TorchToTMTensor.cpp]: do we still need this? for my understanding the limitation before for restricting the scale to `1.0/sqrt(head_dim)` is because we could not pass the scale as a parameter to the op and IREE would always implicitly assume its `1.0/sqrt(head_dim)` if we can set it as attribute now i assume would not need this limitation anymore?

File: lib/Conversion/TorchToTMTensor/TorchToTMTensor.cpp [hunk 2]
```diff
@@ -1934,32 +1935,31 @@ class ConvertAtenScaledDotProductAttentionOp
     // Verify the scale matches the expected 1/sqrt(headDim).
     // See:
     // https://pytorch.org/docs/stable/generated/torch.nn.functional.scaled_dot_product_attention.html
+    mlir::FloatAttr scaleAttr{};
     if (!isa<Torch::NoneType>(scale.getType())) {
       double scaleFloat;
       if (!matchPattern(scale, m_TorchConstantFloat(&scaleFloat)))
```
Comment (ziereis) [inline lib/Conversion/TorchToTMTensor/TorchToTMTensor.cpp:1941]: i guess it would even be nice we could just make the scale a regular input to the op so we don't have to require it to be a constant. However i see how this might be a bit complicated with variadic arg handling the op currently has.

File: lib/Conversion/TorchToTMTensor/TorchToTMTensor.cpp (hunk 1 above)
Comment (AGindinson) [reply lib/Conversion/TorchToTMTensor/TorchToTMTensor.cpp]: Yes, this is a fair point, especially given that among the use-cases are ONNX models with custom scale settings for different version of attention. @zjgarvey what do you think?

File: lib/Conversion/TorchToTMTensor/TorchToTMTensor.cpp (hunk 2 above)
Comment (AGindinson) [reply lib/Conversion/TorchToTMTensor/TorchToTMTensor.cpp:1941]: Yep, agreed... Can there ever be a case in Torch or ONNX though, where a scale would get assigned symbolically? True, it doesn't have to be a constant value, but I'm just struggling to come up with an example.

Comment (ziereis) [reply lib/Conversion/TorchToTMTensor/TorchToTMTensor.cpp:1941]: i think it would be very uncommon in a real model but one example would be passing it in as a function argument which could happen when creating some reproducer or something like this. Either way i don't think its important to address this now, setting it as attribute is already a good improvement over the current state.

**Comment by ziereis** [comment]
looks good from my side. But can't approve formally.

**Review by jtuyls — APPROVED** [review]
LGTM
